src/algebra/euclid-algorithm.md

```diff
@@ -0,0 +1,119 @@
+<!--?title The Euclidean algorithm for finding GCD (greatest common divisor) -->
+
+# The Euclidean algorithm for finding GCD (greatest common divisor)
+
+Given two non-negative integers $a$ and $b$, we require the greater common divisor i.e. the largest number which is a divisor of $a$ and $b$ simultaneously. It's common denoted by $gcd$.
+
+![Euclid’s formula](&imgroot&/euclid1.png)
+
+(here the symbol "$\mid$" denotes divisibility, i.e. “$k\mid a$” means “$k$ divides $a$”)
+
+When one of the numbers is zero, while the other is non zero, the greatest common divisor, by definition, is the second number. However, when both these numbers are zero, it is undefined(it can be an infinitely large number), but we can take it to be equal to zero. So we have a simple rule: if one of the numbers is zero, the other number is the gcd.
+
+The Euclidean algorithm, discussed below, solves the problem of finding the greatest common divisor of two numbers $a$ and $b$ in $O(\log min(a, b))$.
+
+The algorithm was first described in Euclid's “**Elements**” (circa 300 BC). However, it is possible that the algorithm has earlier origins.
+
+## Algorithm
+
+The algorithm is extremely simple and described by this formula:
+
+![image](&imgroot&/euclid2.png)
+
+## Implementation
+
+```cpp
+int gcd (int a, int b) {
+	if (b == 0)
+		return a;
+	else
+		return gcd (b, a % b);
+}
+```
+
+Using the ternary operator in C++, we can write a shorter implementation:
+
+```cpp
+int gcd (int a, int b) {
+	return b ? gcd (b, a % b) : a;
+}
+```
+
+Finally, we present a non-recursive form of implementation of the algorithm:
+
+```cpp
+int gcd (int a, int b) {
+	while (b) {
+ a %= b;
+ swap (a, b);
+	}
+	return a;
+}
+```
+
+## The Proof of correctness
+
+First, observe that at each successive iteration of the Euclidean algorithm, it's second argument strictly decreases, therefore, since it is always non-negative, the algorithm must terminate.
+
+For the proof of correctness, we need to show $gcd(a, b) = gcd(b, a\mod b)$ for all $a \geq 0$, $b > 0$.
+
+We will show that the quantity on the left side of the equation, can be written as the value on the left. Obviously, this would mean that the left and right sides are the same and prove Euclid's algorithm.
```

Comment:
In the original text, the proof shows that the expression of the left side of the equation divides the expression on the right side and vise versa. Let's keep this meaning.